pandas/core/interval.py

```diff
@@ -0,0 +1,521 @@
+        of all intervals:
+
+        >>> Index([Interval(0, 1), Interval(1, 2)])
+        IntervalIndex(left=[0, 1],
+                      right=[1, 2],
+                      closed='right')
+        """
+        data = np.asarray(data)
+        left, right, closed = intervals_to_interval_bounds(data)
+        return cls(left, right, closed, name)
+
+    @classmethod
+    def from_tuples(cls, data, closed='right', name=None):
+        left = []
+        right = []
+        for l, r in data:
+            left.append(l)
+            right.append(r)
+        return cls(np.array(left), np.array(right), closed, name)
+
+    def to_tuples(self):
+        return Index(com._asarray_tuplesafe(zip(self.left, self.right)))
+
+    @cache_readonly
+    def _multiindex(self):
+        return MultiIndex.from_arrays([self.left, self.right],
+                                      names=['left', 'right'])
+
+    @property
+    def left(self):
+        return self._left
+
+    @property
+    def right(self):
+        return self._right
+
+    @property
+    def closed(self):
+        return self._closed
+
+    def __len__(self):
+        return len(self.left)
+
+    @cache_readonly
+    def values(self):
+        """Returns the IntervalIndex's data as a numpy array of Interval
+        objects (with dtype='object')
+        """
+        left = np.asarray(self.left)
+        right = np.asarray(self.right)
+        return interval_bounds_to_intervals(left, right, self.closed)
+
+    def __array__(self, result=None):
+        """ the array interface, return my values """
+        return self.values
+
+    def __array_wrap__(self, result, context=None):
+        # we don't want the superclass implementation
+        return result
+
+    def _array_values(self):
+        return self.values
+
+    def __reduce__(self):
+        return self.__class__, (self.left, self.right, self.closed, self.name)
+
+    def _shallow_copy(self, values=None, name=None):
+        name = name if name is not None else self.name
+        if values is not None:
+            return type(self).from_intervals(values, name=name)
+        else:
+            return self.copy(name=name)
+
+    def copy(self, deep=False, name=None):
+        left = self.left.copy(deep=True) if deep else self.left
+        right = self.right.copy(deep=True) if deep else self.right
+        name = name if name is not None else self.name
+        return type(self)(left, right, closed=self.closed, name=name,
+                          fastpath=True)
+
+    @cache_readonly
+    def dtype(self):
+        return np.dtype('O')
+
+    @cache_readonly
+    def mid(self):
+        """Returns the mid-point of each interval in the index as an array
+        """
+        try:
+            return Index(0.5 * (self.left.values + self.right.values))
+        except TypeError:
+            # datetime safe version
+            delta = self.right.values - self.left.values
+            return Index(self.left.values + 0.5 * delta)
```

Comment:
the other version is faster is it's supported.. but that's probably premature optimization